composer/datasets/webdataset.py

```diff
@@ -0,0 +1,70 @@
+import json
+import os
+import subprocess
+from typing import Any, Dict, Iterable, Optional, Tuple
+
+from tqdm import tqdm
+from webdataset import ShardWriter, WebDataset
+from wurlitzer import pipes
+
+
+def create_webdataset_meta(split_dir: str, n_samples: int, n_shards: int) -> None:
+    '''Write a WebDataset meta file.'''
+    samples_per_shard = n_samples // n_shards
+    n_leftover = n_samples % samples_per_shard
+    obj = {
+        'n_shards': n_shards,
+        'samples_per_shard': samples_per_shard,
+        'n_leftover': n_leftover,
+    }
+    filename = os.path.join(split_dir, 'meta.json')
+    json.dump(obj, open(filename, 'w'), sort_keys=True)
+
+
+def create_webdataset(samples: Iterable[Dict[str, Any]], dataset_dir: str, split: str, n_samples: int,
+                      n_shards: int, use_tqdm: int = 1) -> None:
+    '''Write an entire WebDataset to a local directory, given an iterable of samples.'''
+    split_dir = os.path.join(dataset_dir, split)
+    os.makedirs(split_dir)
+    pattern = os.path.join(split_dir, '%05d.tar')
+    samples_per_shard = n_samples // n_shards
+    with pipes():
+        out = ShardWriter(pattern, maxcount=samples_per_shard)
+        out.verbose = 0
+    if use_tqdm:
+        samples = tqdm(samples, total=n_samples, leave=False)
+    for sample in samples:
+        out.write(sample)
+    out.close()
+    create_webdataset_meta(split_dir, n_samples, n_shards)
+
+
+def download_webdataset_meta(dataset_name: str, split: str) -> bytes:
+    '''Download a WebDataset meta file from S3.'''
+    url = f's3://mosaicml-internal-dataset-{dataset_name}/{split}/meta.json'
```

Comment:
I think this URL input needs to be lifted up to the `DatasetHparams` level, because the bucket path will end up being customer specific.